catalyst-gateway/bin/src/service/common/auth/rbac/scheme.rs

```diff
@@ -172,20 +192,194 @@ async fn checker_api_catalyst_auth(
+            c509_key(cert)
+        },
+        LocalRefInt::PubKeys => {
+            // We check this during Cip509 validation.
+            Err(anyhow!(
+                "Invalid signing key for role 0: it must reference a certificate, not public key"
+            ))
+        },
+    }
+}
+
+/// Build a registration chain from the given indexed data.
+async fn registration_chain(
+    network: Network, indexed_registrations: &[Query],
+) -> anyhow::Result<RegistrationChain> {
+    let mut indexed_registrations = indexed_registrations.iter();
+    let Some(root) = indexed_registrations.next() else {
+        // We already checked that the registrations aren't empty, so we shouldn't get there.
+        return Err(anyhow!("Empty registrations list"));
+    };
+
+    let root = registration(network, root.slot_no.into(), root.txn_index.into())
+        .await
+        .context("Failed to get root registration")?;
+    let mut chain = RegistrationChain::new(root).context("Invalid root registration")?;
+
+    for reg in indexed_registrations {
+        // We only store valid registrations in this table, so an error here indicates a bug in
+        // our indexing logic.
+        let cip509 = registration(network, reg.slot_no.into(), reg.txn_index.into())
+            .await
+            .with_context(|| {
+                format!(
+                    "Invalid or missing registration at {:?} block {:?} transaction",
+                    reg.slot_no, reg.txn_index,
+                )
+            })?;
+        match chain.update(cip509) {
+            Ok(c) => chain = c,
+            Err(e) => {
+                // This isn't a hard error because while the individual registration can be valid it
+                // can be invalid in the context of the whole registration chain.
+                warn!(
+                    "Unable to apply registration from {:?} block {:?} txn index: {e:?}",
+                    reg.slot_no, reg.txn_index
+                );
+            },
+        }
+    }
+
+    Ok(chain)
+}
+
+/// Returns a RBAC registration from the given block and slot.
+async fn registration(network: Network, slot: Slot, txn_index: TxnIndex) -> anyhow::Result<Cip509> {
+    let point = Point::fuzzy(slot);
+    let block = ChainFollower::get_block(network, point)
+        .await
+        .context("Unable to get block")?
+        .data;
+    if block.point().slot_or_default() != slot {
+        // The `ChainFollower::get_block` function can return the next consecutive block if it
+        // cannot find the exact one. This shouldn't happen, but we need to check anyway.
+        return Err(anyhow!("Unable to find exact block"));
+    }
+    Cip509::new(&block, txn_index, &[])
+        .context("Invalid RBAC registration")?
+        .context("No RBAC registration at this block and txn index")
+}
+
+/// Returns `VerifyingKey` from the given X509 certificate.
+fn x509_key(cert: &Certificate) -> anyhow::Result<VerifyingKey> {
+    let oid: Oid = cert
```

Comment:
I assume that this should be subject public key algorithm so it should come from `tbsCertificate` -> `SubjectPublicKeyInfo`